File: demo-node-service/patterns/caching/cacheWarming.js
const EventEmitter = require('events');

class CacheWarmingService extends EventEmitter {
    constructor(cache) {
        super();
        this.cache = cache;
        this.warmingJobs = new Map();
        this.running = true;
    }
    
    scheduleWarming(key, dataLoader, interval = 300000) {
        if (this.warmingJobs.has(key)) {
            clearInterval(this.warmingJobs.get(key));
        }
        
        const warmingJob = setInterval(async () => {
            if (!this.running) return;
            
            try {
                const data = await dataLoader();
                await this.cache.set(key, data);
                this.emit('cacheWarmed', { key, data });
            } catch (error) {
                this.emit('cacheWarmingError', { key, error: error.message });
            }
        }, interval);
        
        this.warmingJobs.set(key, warmingJob);
    }
    
    async warmCache(cacheSpecs) {
        const promises = cacheSpecs.map(async ({ key, dataLoader }) => {
            try {
                const data = await dataLoader();
                await this.cache.set(key, data);
                return { key, success: true };
            } catch (error) {
                return { key, success: false, error: error.message };
            }
        });
        
        const results = await Promise.all(promises);
        this.emit('batchWarmingCompleted', results);
        return results;
    }
    
    stopWarming(key) {
        if (this.warmingJobs.has(key)) {
            clearInterval(this.warmingJobs.get(key));
            this.warmingJobs.delete(key);
        }
    }
    
    stopAllWarming() {
        this.running = false;
        for (const [key, job] of this.warmingJobs) {
            clearInterval(job);
        }
        this.warmingJobs.clear();
    }
    
    getWarmingStats() {
        return {
            activeJobs: this.warmingJobs.size,
            running: this.running
        };
    }
}

class CacheInvalidationService extends EventEmitter {
    constructor(cache) {
        super();
        this.cache = cache;
        this.patterns = new Map();
    }
    
    registerPattern(patternName, keyPattern) {
        this.patterns.set(patternName, new RegExp(keyPattern));
    }
    
    async invalidateByPattern(patternName) {
        const pattern = this.patterns.get(patternName);
        if (!pattern) {
            throw new Error(`Pattern ${patternName} not registered`);
        }
        
        const keysToInvalidate = [];
        
        try {
            const allKeys = await this.cache.getAllKeys();
            for (const key of allKeys) {
                if (pattern.test(key)) {
                    keysToInvalidate.push(key);
                }
            }
        } catch (error) {
            // Fallback if cache doesn't support getAllKeys
            console.warn('Cache does not support getAllKeys, pattern invalidation limited');
        }
        
        const results = await Promise.all(
            keysToInvalidate.map(async (key) => {
                try {
                    await this.cache.delete(key);
                    return { key, success: true };
                } catch (error) {
                    return { key, success: false, error: error.message };
                }
            })
        );
        
        this.emit('patternInvalidated', { pattern: patternName, results });
        return results;
    }
    
    async invalidateByPrefix(prefix) {
        const keysToInvalidate = [];
        
        try {
            const allKeys = await this.cache.getAllKeys();
            for (const key of allKeys) {
                if (key.startsWith(prefix)) {
                    keysToInvalidate.push(key);
                }
            }
        } catch (error) {
            console.warn('Cache does not support getAllKeys, prefix invalidation limited');
        }
        
        const results = await Promise.all(
            keysToInvalidate.map(async (key) => {
                try {
                    await this.cache.delete(key);
                    return { key, success: true };
                } catch (error) {
                    return { key, success: false, error: error.message };
                }
            })
        );
        
        this.emit('prefixInvalidated', { prefix, results });
        return results;
    }
    
    getInvalidationStats() {
        return {
            registeredPatterns: Array.from(this.patterns.keys())
        };
    }
}

module.exports = {
    CacheWarmingService,
    CacheInvalidationService
};

File: demo-python-service/patterns/caching/cache_warming.py
import threading
import time
from typing import Dict, Any, Callable, List
import logging

logger = logging.getLogger(__name__)

class CacheWarmingService:
    def __init__(self, cache):
        self.cache = cache
        self.warming_jobs = {}
        self.running = True
        
    def schedule_warming(self, key: str, data_loader: Callable, interval: int = 300):
        """Schedule periodic cache warming for a key"""
        def warming_worker():
            while key in self.warming_jobs and self.running:
                try:
                    data = data_loader()
                    self.cache.set(key, data)
                    logger.info(f"Cache warmed for key: {key}")
                except Exception as e:
                    logger.error(f"Cache warming failed for {key}: {e}")
                time.sleep(interval)
                
        thread = threading.Thread(target=warming_worker, daemon=True)
        self.warming_jobs[key] = thread
        thread.start()
        
    def warm_cache(self, cache_specs: List[Dict[str, Any]]):
        """Warm cache with multiple keys at once"""
        for spec in cache_specs:
            key = spec['key']
            data_loader = spec['data_loader']
            try:
                data = data_loader()
                self.cache.set(key, data)
                logger.info(f"Cache warmed for key: {key}")
            except Exception as e:
                logger.error(f"Failed to warm cache for {key}: {e}")
                
    def stop_warming(self, key: str):
        """Stop warming for a specific key"""
        if key in self.warming_jobs:
            del self.warming_jobs[key]
            
    def stop_all_warming(self):
        """Stop all warming jobs"""
        self.running = False
        self.warming_jobs.clear()

class CacheInvalidationService:
    def __init__(self, cache):
        self.cache = cache
        self.patterns = {}
        
    def register_pattern(self, pattern_name: str, key_pattern: str):
        """Register a pattern for cache invalidation"""
        self.patterns[pattern_name] = key_pattern
        
    def invalidate_by_pattern(self, pattern_name: str):
        """Invalidate cache entries matching a pattern"""
        if pattern_name not in self.patterns:
            return
            
        import re
        pattern = self.patterns[pattern_name]
        keys_to_invalidate = []
        
        # Get all cache keys and match pattern
        try:
            for key in self.cache.get_all_keys():
                if re.match(pattern, key):
                    keys_to_invalidate.append(key)
        except AttributeError:
            # Fallback if cache doesn't have get_all_keys
            logger.warning("Cache doesn't support get_all_keys, pattern invalidation limited")
            
        for key in keys_to_invalidate:
            self.cache.delete(key)
            logger.info(f"Invalidated cache key: {key}")
            
    def invalidate_by_prefix(self, prefix: str):
        """Invalidate all cache entries with a specific prefix"""
        keys_to_invalidate = []
        try:
            for key in self.cache.get_all_keys():
                if key.startswith(prefix):
                    keys_to_invalidate.append(key)
        except AttributeError:
            logger.warning("Cache doesn't support get_all_keys, prefix invalidation limited")
            
        for key in keys_to_invalidate:
            self.cache.delete(key)
            logger.info(f"Invalidated cache key: {key}")